### PR TITLE
Experimental Zen3 scheduler

### PR DESCRIPTION
**Pull request by kd-11**
- Test if forcing sample shading rate fixes navi borked MSAA
- Test if thread affinity is broken (zen3 test)

### CODE REVIEW

**Comment by hyper92** [comment]
Is working better for me then master @5900x in RDR and other games.
Before this I had random freezes pretty often (very often already in the title screen of the game) especially in RDR.. now they seem to be all gone. Framerate is also a bit better than current master as well..

Will try out some more games tomorrow. Is there some specific info you would need?

**Comment by Reonu** [comment]
Runs better on my 5800X, but not by much:

Master:
imagen

PR:
imagen

The biggest improvement is seen from disabling SMT and running the 10736 build:
imagen

My GPU is a RTX 3070. My 5800X runs at around 4.7 GHz when running RPCS3.

**Comment by JohnHolmesII** [comment]
You finally did it. Madlad.

Master:
master

PR:
pr

**Comment by AniLeo** [comment]
It's still not fully fixed, but yes, wireframe itself is gone, however another issue now happens on games such as Skate 3 and RDR, we're debugging with RDNA2 as I finally got a GPU

**Comment by JohnHolmesII** [comment]
Would you care to *share* what it is? I just started testing RDR as well.

**Comment by AniLeo** [comment]
Was already editing right after posting, had to look for the screenshot

Regarding Skate 3, it's not very noticeable but look below the skate

image

**Comment by JohnHolmesII** [comment]
Might be a NAVI2 thing:
image

**Comment by kd-11** [comment]
Maybe we have two different bugs on navi 2x. Looks like the workaround fixes rendering on navi1x completely which is unexpected. We did uncover some unexpected spec violations that I am looking into but its good that the original cards are ok with the fix.

**Comment by AniLeo** [comment]
Another note: The wireframe was never visible when RenderDoc was injected into RPCS3, only that depth corruption from the screenshot above. With this PR, RPCS3 no longer displays the wireframe but displays what we would see on RenderDoc prior to this PR.

**Comment by JohnHolmesII** [comment]
There is an odd bug that appears when enabling msaa that occurs. Seems like transparencies flicker oddly. Seems dependent on camera angle. I know windows and such have been an issue in RDR since it first went ingame lol, but this looks at least a bit more digestible to fix.

Video is available [here](http://66.42.125.146/misc/navi_experiments_flickering.mkv)

**Comment by Nekotekina** [comment]
Wasn't you supposed to disable "thread scheduler" setting?
Otherwise, I use thread pinning in TSC calibration code.

**Comment by Nekotekina** [comment]
Couldn't find any other use which was not under thread scheduler setting.

**Comment by kd-11** [comment]
That was a cheeky placebo test to eliminate some dubious results. I'm adding correct scheduler for zen3 correctly.

**Comment by kd-11** [comment]
@Reonu You can retest once the build is ready. Just check with SMT enabled and thread scheduler option also selected.

**Comment by Reonu** [comment]
@kd-11 New version of the PR with SMT ON and thread scheduler ON
imagen

I can't test again until friday since I'm gonna be at uni

**Comment by kd-11** [comment]
NAVI MSAA patch has been added to master fixing the wireframe problem. Stripes on RDNA2 still require investigation.

**Comment by AniLeo** [comment]
RNDA2 bug fixed by 20.11.2 VRT Beta driver
https://www.amd.com/en/support/kb/release-notes/rn-rad-win-20-11-2-vrt-beta

Tested on RX 6800 XT

**Comment by aqxa1** [comment]
I just tested out the thread scheduler changes, and it's tanked performance on my 5900X (on Linux).  Here's with the WIP scheduler:
WIP scheduler

And the 3900x scheduler:
3900x scheduler

Just looking at the comment/code here:
```
// Single-CCX architecture, just disable SMT if wide enough
// CCX now holds upto 16 threads
if (thread_count >= 16)
```

Is this assuming that all Zen 3 CPUs that have 16 or greater threads have an 8 core/16 thread CCX? Because that's not true for the 5900x, which is 2x (6/12 CCXs) cut down from the full 2x 8/16 of the 5950x. Which might explain the performance drop. I've also tested limiting to one CCX with `taskset` and that also gives very good performance, although probably a bit worse than the 3900x scheduler.

Of course, perhaps this configuration will work better on some CPU schedulers (perhaps on Windows). I'll have to test MuQSS again, but it gives pretty horrendous performance on the 5900x by default (even with the thread scheduler disabled) which is strongly in contrast to the 2700 where it performed extremely well.

File: Utilities/Thread.cpp [hunk 1]
```diff
@@ -2434,105 +2434,129 @@ u64 thread_ctrl::get_affinity_mask(thread_class group)
...
+				if (model_id > 0x30)
 				{
-					// Ryzen 7 3700/3800 (x)
-					// Assign threads 1-16
-					ppu_mask = 0b0000000011110000;
-					spu_mask = 0b1111111100000000;
-					rsx_mask = 0b0000000000001111;
+					// Zen2 (models 49, 96, 113, 144)
+					// Much improved inter-CCX latency
+					switch (thread_count)
```
Comment (AniLeo) [inline Utilities/Thread.cpp:2459]: Duplicated switch

**Comment by kd-11** [comment]
I have made 5900X use the 3900X scheduler. Lack of access to these chips is making it difficult to pick the optimal strategy, I would theoretically need a 5950X to simulate all possible configurations using downcore. I think I will merge this in a few days if it doesn't hurt performance. We can update the code down the line as more CPUs become available.

**Comment by Reonu** [comment]
> 
> 
> I have made 5900X use the 3900X scheduler. Lack of access to these chips is making it difficult to pick the optimal strategy, I would theoretically need a 5950X to simulate all possible configurations using downcore. I think I will merge this in a few days if it doesn't hurt performance. We can update the code down the line as more CPUs become available.

What about the 5800X? It should be simple for that one, as you can simply use any / all threads since there's only one CCX

File: Utilities/Thread.cpp [hunk 2]
```diff
@@ -2493,105 +2493,140 @@ u64 thread_ctrl::get_affinity_mask(thread_class group)
...
+					case 64:
+					case 48:
+					case 32:
+						// TR 3000 series, or R9 3950X, Assign threads 9-32
+						ppu_mask = 0b11111111000000000000000000000000;
+						spu_mask = 0b00000000111111110000000000000000;
+						rsx_mask = 0b00000000000000001111111100000000;
+						break;
+					case 24:
+						// 3900X, Assign threads 7-22
```
Comment (sampletext32) [inline Utilities/Thread.cpp:2530]: incorrect comment judjing by the code
```suggestion
						// 3900X, Assign threads 7-24
```

**Comment by kd-11** [comment]
> > I have made 5900X use the 3900X scheduler. Lack of access to these chips is making it difficult to pick the optimal strategy, I would theoretically need a 5950X to simulate all possible configurations using downcore. I think I will merge this in a few days if it doesn't hurt performance. We can update the code down the line as more CPUs become available.
> 
> What about the 5800X? It should be simple for that one, as you can simply use any / all threads since there's only one CCX

5800X and smaller are handled.

File: Utilities/Thread.cpp [hunk 3]
```diff
@@ -2493,105 +2493,140 @@ u64 thread_ctrl::get_affinity_mask(thread_class group)
...
+						spu_mask = ppu_mask;
+						rsx_mask = 0b11111111000000000000000000000000;
+						break;
+					case 48:
+						// TR 2970WX, Assign threads 9-24
+						ppu_mask = 0b000000111111000000000000;
+						spu_mask = ppu_mask;
+						rsx_mask = 0b111111000000000000000000;
+						break;
+					case 32:
```
Comment (arabek) [inline Utilities/Thread.cpp:2568]: 1950X under Linux assigns:
$ cat /sys/devices/system/node/node0/cpulist
0-7,16-23
$ cat /sys/devices/system/node/node1/cpulist
8-15,24-31
to numa node 0 and 1 respectively. w/r to where the gpu is located, numa node1 (and cpu node1) are nearer (local) to the gpu.

```
  Group0 L#1
    NUMANode L#1 (P#1 31GB)
    L3 L#2 (8192KB)
      L2 L#8 (512KB) + L1d L#8 (32KB) + L1i L#8 (64KB) + Core L#8
        PU L#16 (P#8)
        PU L#17 (P#24)
      L2 L#9 (512KB) + L1d L#9 (32KB) + L1i L#9 (64KB) + Core L#9
        PU L#18 (P#9)
        PU L#19 (P#25)
      L2 L#10 (512KB) + L1d L#10 (32KB) + L1i L#10 (64KB) + Core L#10
        PU L#20 (P#10)
        PU L#21 (P#26)
      L2 L#11 (512KB) + L1d L#11 (32KB) + L1i L#11 (64KB) + Core L#11
        PU L#22 (P#11)
        PU L#23 (P#27)
    L3 L#3 (8192KB)
      L2 L#12 (512KB) + L1d L#12 (32KB) + L1i L#12 (64KB) + Core L#12
        PU L#24 (P#12)
        PU L#25 (P#28)
      L2 L#13 (512KB) + L1d L#13 (32KB) + L1i L#13 (64KB) + Core L#13
        PU L#26 (P#13)
        PU L#27 (P#29)
      L2 L#14 (512KB) + L1d L#14 (32KB) + L1i L#14 (64KB) + Core L#14
        PU L#28 (P#14)
        PU L#29 (P#30)
      L2 L#15 (512KB) + L1d L#15 (32KB) + L1i L#15 (64KB) + Core L#15
        PU L#30 (P#15)
        PU L#31 (P#31)
    HostBridge
      PCIBridge
        PCI 41:00.0 (VGA)
          GPU ":0.0"
          CoProc(CUDA) "cuda0"
          GPU "nvml0"
      PCIBridge
        PCI 43:00.2 (SATA)
```

Hope it helps

**Comment by fifazalata** [comment]
I have 5900x on Window 10.
I've been getting slightly better performance with this PR than Master but only with Thread Scheduler OFF
When turned ON, I get a much worse performance.

Example:
Thread Scheduler OFF:


Thread Scheduler ON:


Logs:
[RPCS3.log-Scheduler OFF.gz](https://github.com/RPCS3/rpcs3/files/5682995/RPCS3.log-Scheduler.OFF.gz)
[RPCS3.log-Scheduler ON.gz](https://github.com/RPCS3/rpcs3/files/5682988/RPCS3.log-Scheduler.ON.gz)
:

**Comment by Asinin3** [comment]
What do you get in master with scheduler off?

**Comment by fifazalata** [comment]
It's merely a 1-3 fps difference increase at most than Master (if even there is an actual increase)
But Ok. I'll do it & update the original comment.

File: Utilities/Thread.cpp (hunk 3 above)
Comment (Hirato) [reply Utilities/Thread.cpp:2568]: I also have a 1950X as well.
I've found the layout of the CPU cores changes radically between BIOS revisions.
I think some of them grouped the SMT threads together, so node 0 and 1 would be 0-15, 16-31 respectively.
for AGESA 1.1.0.2 however, it should be 0-7,16-23/8-15,24-31 as noted above.

If you have a NUMA threadripper system, I would suggest you just forget about the scheduler, and run the emu with numactl instead,
i.e. `numactl -N 1 --localalloc rpcs3`

**Comment by Zeadus** [comment]
This makes Demon souls playable for me using a 5600x on W10, i was able to clear the game without a single crash at stable 60fps. Without this FPS would start at 60fps and then go down to 3-4fps and get stuck there.

CPU Usage was also at 30%~ the whole game.

**Comment by kd-11** [comment]
I'm still not sure if this PR is beneficial or not. There just aren't enough zen3 testers to confirm if this is worth committing right now.

**Comment by fifazalata** [comment]
If you need more tests in the future, let me know.

**Comment by DefaltBR** [comment]
I got a 5600X to help testing... But i can't remember how to get the PR build to test... X.x

**Comment by Reonu** [comment]
I can do more tests on my 5800X if needed, just tell me what to do

**Comment by Asinin3** [comment]
Need to wait for kd-11 to rebase this pull request. Latest commit shows it wasn't built for windows. Previous commit doesn't show azure pipeline built. I'll do some extensive testing with my 5950x when it's rebased.

**Comment by Nekotekina** [comment]
Solved conflicts, apparently it was my cleanup that prevented easy rebase anyway.

**Comment by Nekotekina** [comment]
https://dev.azure.com/nekotekina/nekotekina/_build/results?buildId=5721&view=artifacts&pathAsName=false&type=publishedArtifacts I think artifacts for testing can be downloaded from here

**Comment by Asinin3** [comment]
5950X Benches, AVG FPS taken with afterburner. Stationary for 1-2mins

GoW 3 - City of Olympus, after cutscene in cave exit out to city
PR Scheduler On: 44.3FPS AVG - Not a fluke, mins/max were higher also.
PR Scheduler off: 43.5FPS AVG
Master Scheduler On: 32 FPS - Scheduler is awful on master with a 5950x, was good on my old 3700x though.
Master Scheduler Off: 43 FPS - Min was definitely lower here.

**Comment by fifazalata** [comment]
Tested Red Dead Redemption with 5900X, nearly same results as @Asinin3
merely 1-2 fps (if any) increase when PR Scheduler On

**Comment by Satan86** [comment]
Few tests with my 5600x:
Gran Turismo 5 - PR Thread Scheduler OFF: 70-75 FPS
Gran Turismo 5 - PR Thread Scheduler ON: 68-72 FPS

RDR - PR Thread Scheduler OFF: 24-26 FPS*
RDR - PR Thread Scheduler ON: 23-25 FPS
*RDR FPS varies a lot between different runs, so I made sure to run the game a few times with Scheduler on and off. I only got that FPS once, other runs were the same FPS on both options.

Uncharted 1 - PR Scheduler OFF : 33-34 FPS
Uncharted 1 - PR Scheduler ON: 33-35 FPS

Performance and stability with thread scheduler on the latest master are terrible

**Comment by kd-11** [comment]
In that case I'll merge this as-is. It seems to at least match scheduler OFF or get close enough without causing serious performance degradation. This can be revisited in future once there is more hardware in circulation I think, as linux still offers up the optimal performance for zen chips.

**Comment by Modey2222** [comment]
i have a 5800x i can be your tester but i dont know how to contact u guys
but performance is so bad compared to intel

**Comment by DefaltBR** [comment]
The new Thread Scheduler implementation is causing random microfreezes/stutter for me in almost every game i enable it, it only works fine if Accurate RSX Reservation Access is used along though... 5600X here.

**Comment by Tetriando** [comment]
@kd-11 
5800x user here, I just checked Demon's Souls with latest version and unfortunately stutters make it unplayable.
I would like to help with testing when this topic will be revisited in the future.

**Comment by Tetriando** [comment]
@kd-11 
I checked again with latest version of RPCS3 today and it was the same but after setting thread scheduler to OFF it completely solved my issues with Demon's Souls.

What I noticed with thread scheduler ON is that constantly clicking mouse button while using controller to play also helped (no idea how that is connected but maybe it's a hint).

**Comment by Asinin3** [comment]
It's pretty simple, if you get better performance with the scheduler off, then turn it off. My 5950x only performs better with it on if I have SMT enabled. This CPU has two CCX's unlike your 5800x, which should mean it's less likely to need the scheduler to begin with.

**Pull request merged**